node/modules/mpoolnonceapi.go

```diff
@@ -21,13 +22,41 @@ type MpoolNonceAPI struct {
 	StateAPI full.StateAPI
 }
 
-// GetNonce gets the nonce from actor state
+// GetNonce gets the nonce from current chain head.
 func (a *MpoolNonceAPI) GetNonce(addr address.Address) (uint64, error) {
-	act, err := a.StateAPI.StateGetActor(context.Background(), addr, types.EmptyTSK)
+	ts := a.StateAPI.Chain.GetHeaviestTipSet()
+
+	// make sure we have a key address so we can compare with messages
+	keyAddr, err := a.StateAPI.StateManager.ResolveToKeyAddress(context.TODO(), addr, ts)
+	if err != nil {
+		return 0, err
+	}
+
+	// Load the last nonce from the state, if it exists.
+	highestNonce := uint64(0)
+	if baseActor, err := a.StateAPI.StateManager.LoadActorRaw(context.TODO(), addr, ts.ParentState()); err != nil {
+		if !xerrors.Is(err, types.ErrActorNotFound) {
+			return 0, err
+		}
+	} else {
+		highestNonce = baseActor.Nonce
+	}
+
+	// Otherwise, find the highest nonce in the tipset.
+	msgs, err := a.StateAPI.Chain.MessagesForTipset(ts)
 	if err != nil {
 		return 0, err
 	}
-	return act.Nonce, nil
+	for _, msg := range msgs {
+		vmmsg := msg.VMMessage()
+		if vmmsg.From != keyAddr {
```

Comment:
I think msg.From can technically also be an ID address, though I don't see messages like that on-chain

Reply:
You sure? It looks like message verification will fail if the from address is not a public key.

https://github.com/filecoin-project/lotus/blob/4937fb925e4327831993efbe9c6d47c653f1603e/chain/messagepool/messagepool.go#L635-L637

Reply:
Uff we should fix it.

Reply:
yeah, you should abosolutely be able to send a message with an ID address as the from

Reply:
Well, we can add that to the network upgrade. Note: we'll need to carefully check the VM to make sure we resolve everything as expected before passing it off to the actors.

Reply:
Note: we currently _implicitly_ resolve the from address when sending. Should we not be doing that?